### PR TITLE
Improve error handling

### LINKED ISSUE #138: Show or prevent submit if amount > balance
Would be good to
 - show the users token balance, maybe have it inside the asset selector card? Or next to the amount input
Could be one of these places perhaps. 
image

- show an error + input red/warning state when the amount input is > than the users balance.

_Originally posted by @0xjimmy in https://github.com/DarkFlorist/lunaria/pull/133#pullrequestreview-1461441866_

### PR DESCRIPTION
**Pull request by jubalm**
Improve error handling, adding proper notices and error messages on transfer:
- [x] user rejection as a quick toast message instead of error
- [x] handle RPC errors unbeknownst to ethersjs
- [x] refactor ethereum provider details, moving it to `Ethereum` context
- [x] refactor account details, moving it to `Wallet` context
- [x] remove unused files and functions

Notable UI update:

Screenshot 2023-10-10 at 4 00 20 PM

Resolves #138

### CODE REVIEW

File: app/ts/components/TokenPicker.tsx [hunk 1]
```diff
@@ -220,12 +230,17 @@ const RemoveAssetDialog = ({ token }: { token: ERC20Token }) => {
...
+const AddTokenOrConnectCard = () => {
+	const { account } = useWallet()
 	const { stage } = useTokenManager()
 	const openAddTokenDialog = () => {
 		stage.value = 'add'
 	}
 
+	if (account.value.state !== 'resolved') return (
+		<></>
+	)
```
Comment (KillariDev) [inline app/ts/components/TokenPicker.tsx:242]:
```suggestion
	if (account.value.state !== 'resolved') return <></>
```

File: app/ts/components/TransactionPage/TransactionDetails.tsx [hunk 2]
```diff
@@ -155,28 +154,7 @@ type TokenAmountProps = {
...
-			return <InfoError displayText='Failed to get token amount' message={query.value.error.message} />
-		case 'resolved':
-			const { decimals, symbol } = query.value.value
-			const amount = formatUnits(tokenValue, decimals)
-			addFavoriteStore.value = { ...addFavoriteStore.peek(), amount, token: query.value.value }
-			return <Info label='Amount' value={`${amount} ${symbol}`} />
-	}
+	// TODO: implement amount query
+	console.log('token amount', receipt, addFavoriteStore)
+	return <></>
```
Comment (KillariDev) [inline app/ts/components/TransactionPage/TransactionDetails.tsx:159]: I feel this should be done before this is merged?

File: app/ts/components/TransferButton.tsx [hunk 3]
```diff
@@ -27,17 +27,15 @@ export const TransferButton = () => {
...
 
 	const connect = () => {
-		if (!browserProvider) {
-			notify({ message: 'No compatible web3 wallet detected.', title: 'Failed to connect' })
-			return
-		}
 		waitFor(async () => {
-			const signer = await browserProvider.getSigner()
+			if (!browserProvider.value) {
+				throw makeError('No compatible web3 wallet detected.', 'UNKNOWN_ERROR', { error: { code: 4900 } })
```
Comment (KillariDev) [inline app/ts/components/TransferButton.tsx:36]: this error message and code repeats, we could make it a constant or a function

File: app/ts/library/utilities.ts [hunk 4]
```diff
@@ -58,3 +56,11 @@ export function stringIncludes(source: string, search: string, caseSensitive?: b
 export function preventFocus(e: JSX.TargetedEvent<HTMLElement>) {
 	e.currentTarget.blur()
 }
+
+/**
+ * Match string equality
+ */
+export function areEqualStrings(a: string, b: string, caseSensitive?: true) {
```
Comment (KillariDev) [inline app/ts/library/utilities.ts:63]: do you need non-case sensitive check? if so, you could just not call this function and use "a === b". And when you do, you could rename this function to "areEqualStringsCaseInsentive" or similar

File: app/ts/context/Notification.tsx [hunk 5]
```diff
@@ -0,0 +1,96 @@
+import { ComponentChildren, createContext } from "preact"
+import { Signal, useSignal, useSignalEffect } from "@preact/signals"
+import { useContext, useEffect } from "preact/hooks"
+import { useSignalRef } from "../library/preact-utilities.js"
```
Comment (KillariDev) [inline app/ts/context/Notification.tsx:4]:
```suggestion
import { ComponentChildren, createContext } from 'preact'
import { Signal, useSignal, useSignalEffect } from '@preact/signals'
import { useContext, useEffect } from 'preact/hooks'
import { useSignalRef } from '../library/preact-utilities.js'
```

File: app/ts/library/errors.ts [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+		default:
+			console.error('Found unknown error: ', error)
+			return { warning: true, message: `Unknown Error: ${typeof error === 'string' ? error : JSON.stringify(error)}` }
+	}
+}
+
+/*
+ * https://eips.ethereum.org/EIPS/eip-1474#error-codes
+ * */
+const jsonRpcErrorCodes = [-32002,4100,4900] as const
```
Comment (KillariDev) [inline app/ts/library/errors.ts:102]:
```suggestion
const jsonRpcErrorCodes = [-32002, 4100, 4900] as const
```

**Review by KillariDev — APPROVED** [review]

File: app/ts/components/TransferButton.tsx (hunk 3 above)
Comment (jubalm) [reply app/ts/components/TransferButton.tsx:36]: True, created an issue for this #212

File: app/ts/components/TransactionPage/TransactionDetails.tsx (hunk 2 above)
Comment (jubalm) [reply app/ts/components/TransactionPage/TransactionDetails.tsx]: updated the current PR to include the changes

File: app/ts/library/utilities.ts (hunk 4 above)
Comment (jubalm) [reply app/ts/library/utilities.ts:63]: Tracking this suggestion to #213 to push needed functionality to main

**Pull request merged**